--- 
params:
    samp_data: "_test_data/metadata_medium.csv"
    ref_data: "_test_data/merged_assembly_stats.tsv"
    snp_phylos: "_test_data/xan_test_22_331_assembly.treefile"
    snp_align: "_test_data/xan_test_22_331_assembly.fasta"
    ani_matrix: "_test_data/comp.csv"
    core_phylo: "_test_data/xan_test.treefile"
    output_dir: "_test_data/output"
---

# Diversity

How do the samples compare to eachother?

```{r div_setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.width = 10, warning = FALSE)
```

```{r div_libraries}
library(phylocanvas)
library(ape)
```

```{r div_parse_inputs}
ref_meta <- read.csv(params$ref_data, sep = '\t')
ref_meta$modified_id <- gsub(ref_meta$LastMajorReleaseAccession, pattern = ".", replacement = "_", fixed = TRUE)
samp_meta <- read.csv(params$samp_data, sep = ',')
samp_meta$modified_id <- paste0(gsub(samp_meta$sample, pattern = "-", replacement = "_", fixed = TRUE), "_T1")
ani_matrix <-  read.csv(params$ani_matrix, sep = ',', check.names = FALSE)
core_tree <- ape::read.tree(params$core_phylo)
snp_trees <- ape::read.tree(params$snp_phylos)
```

## SNP phylogeny

```{r div_snp_phylo, fig.height = 7, eval = ! is.null(snp_trees)}
# Root tree 
snp_trees <- root(snp_trees, "REF")

# Plot tree
phycanv <- phylocanvas(snp_trees, treetype = "rectangular", alignlabels = T, showscalebar = T, width = "100%")
phycanv
```

```{asis div_no_snp_phylo, echo = is.null(snp_trees)}
There is no tree to draw, probably because there were too few samples.
More info will be added later.
```


## Minimum spanning network

