--- 
params:
    samp_data: "_test_data/metadata_medium.csv"
    ref_data: "_test_data/merged_assembly_stats.tsv"
    snp_phylos: "_test_data/xan_test_22_331_assembly.treefile"
    snp_align: "_test_data/xan_test_22_331_assembly.fasta"
    ani_matrix: "_test_data/comp.csv"
    core_phylo: "_test_data/xan_test.treefile"
    output_dir: "_test_data/output"
---

# Quality control

* A quick indicator of status of each step. 

## Input data quality

* multiqc link

## Downloaded references

* quast
* table with rows for each sample with info on references chosen
   - sample id
   - reference id
   - ANI between sample and reference
* table with one row per reference (taxon id, GSA id, classification, link to ncbi)
* sourmash output (tree?)

## Assembly and annotation

* depth of coverage
* quast link
* BUSCO gene content?
* bakta output?       

## Variant calling?

* vcfr for plots
* iqtree model selection, number of informative sites, indels

## Core genome phylogeny

* core gene info (how many genes, length, paralogs)
* outlier samples causing few genes to be chosen
* iqtree model selection, number of informative sites, indels