--- 
params:
    group: "xan_test"
    refs: "22_331_assembly"
    samp_data: "_test_data/metadata_medium.csv"
    ref_data: "_test_data/merged_assembly_stats.tsv"
    sendsketch: "_test_data/sendsketch"
    variant_data: "_test_data/variant_data"
    ani_matrix: "_test_data/xan_test_comp.csv"
    core_phylo: "_test_data/xan_test.treefile"
    multiqc: "_test_data/multiqc"
    quast: "_test_data/quast"
    versions: "_test_data/software_versions.yml"
execute:
  echo: false
---

# Quality control

```{r knitr_settings, file='knitr_settings.R'}
```

```{r load_libraries, warning=FALSE, message=FALSE, file='load_libraries.R'}
# When running manually for development: source('load_libraries.R')
```

```{r parse_inputs, file='parse_inputs.R'}
# When running manually for development: source('parse_inputs.R')
```

* A quick indicator of status of each step. 

## Input data quality

* multiqc link

## Downloaded references

* quast
* table with rows for each sample with info on references chosen
   - sample id
   - reference id
   - ANI between sample and reference
* table with one row per reference (taxon id, GSA id, classification, link to ncbi)
* sourmash output (tree?)

## Assembly and annotation

* depth of coverage
* quast link
* BUSCO gene content?
* bakta output?       

## Variant calling?

* vcfr for plots
* iqtree model selection, number of informative sites, indels

## Core genome phylogeny

* core gene info (how many genes, length, paralogs)
* outlier samples causing few genes to be chosen
* iqtree model selection, number of informative sites, indels