---
title: "Test report for variant calling"
output: html_document
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  fasta_align: NULL
  metadata: NULL
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prepare

```{r}
# install.packages("adegenet")
# install.packages("poppr")
# library(ade4)
library(adegenet)
library(poppr)
# library(igraph)
library(ape)
library(readr)
```


Load metadata

```{r}
metadata <- read_tsv(params$metadata)
print(metadata)
```

Load fasta alignments

```{r}
pangenome <- read.dna(params$fasta_align, format = "fasta") #Graphtyper
pangenome <- pangenome[which(rownames(pangenome) != "REF"),] #remove REF seq
pangenome.gi <- DNAbin2genind(pangenome)
```


## Plot 

```{r}
pgpist <- as.genclone(pangenome.gi)
mlg.filter(pgpist, distance = bitwise.dist, algorithm = "average_neighbor",  percent = FALSE) <- 3 
ms.loc <- poppr.msn(pgpist,
                    distmat = bitwise.dist(pgpist, percent = FALSE),
                    include.ties = TRUE,
                    mlg=FALSE)
```

